File: java/src/main/java/problem03/Problem03.java
package problem03;

import java.util.LinkedList;
import java.util.List;

/**
 * This problem was asked by Google.
 *
 * Given the root to a binary tree, implement serialize(root),
 * which serializes the tree into a string, and deserialize(s),
 * which deserializes the string back into the tree.
 *
 * For example, given the following Node class
 *
 * class Node:
 *     def __init__(self, val, left=None, right=None):
 *         self.val = val
 *         self.left = left
 *         self.right = right
 * The following test should pass:
 *
 * node = Node('root', Node('left', Node('left.left')), Node('right'))
 * assert deserialize(serialize(node)).left.left.val == 'left.left'
 */
class Problem03 {

    static String serialize(Node node) {
        if (node == null) {
            return "";
        }
        return "["+node.val+","+serialize(node.left)+","+serialize(node.right)+"]";
    }

    static Node deserialize(String text) {
        return deserialize(text, 1);
    }

    private static Node deserialize(String text, int currLevel) {
        if (text.equals("")) {
            return null;
        } else if (text.charAt(0) != '[' || text.charAt(text.length() - 1) != ']') {
            throw new RuntimeException("Malformed tree");
        }
        int level = currLevel - 1;
        List<String> parts = new LinkedList<>();
        StringBuilder part = new StringBuilder();
        for (char c : text.toCharArray()) {
            if (c == '[') {
                level++;
                if (currLevel != level) {
                    part.append(c);
                }
            } else if (c == ',') {
                if (currLevel == level) {
                    parts.add(part.toString());
                    part = new StringBuilder();
                } else {
                    part.append(c);
                }
            } else if (c == ']') {
                if (currLevel == level) {
                    parts.add(part.toString());
                } else {
                    part.append(c);
                }
                level--;
            } else {
                part.append(c);
            }
        }
        if (parts.size() != 3) {
            throw new RuntimeException("Malformed tree");
        }
        return new Node(parts.get(0), deserialize(parts.get(1), currLevel+1), deserialize(parts.get(2), currLevel+1));
    }
}

File: python/problem03.py
class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def serialize(node):
	if node == None:
		return ''
	return '['+node.val+','+serialize(node.left)+','+serialize(node.right)+']'

def deserialize(text):
	return innerDeserialize(text, 1)

def innerDeserialize(text, curLevel):
	if text == '':
		return None
	elif not text.startswith('[') and not text.endswith(']'):
		print('Malformed text')
		return None
	level = curLevel-1
	parts = []
	strPart = ''
	for i in range(len(text)):
		if text[i] == '[':
			level = level+1
			if curLevel != level:
				strPart = strPart+text[i]
		elif text[i] == ',':
			if curLevel == level:
				parts.append(strPart)
				strPart = ''
			else:
				strPart = strPart+text[i]
		elif text[i] == ']':
			if level == curLevel:
				parts.append(strPart)
			else:
				strPart = strPart+text[i]
			level = level-1
		else:
			strPart = strPart+text[i]
	if len(parts) != 3:
		print("Invalid tree")
		return None
	return Node(parts[0], innerDeserialize(parts[1], curLevel+1), innerDeserialize(parts[2], curLevel+1))

node = Node('root', Node('left', Node('left.left')), Node('right'))
text = serialize(node)
print(text)
print(serialize(deserialize(text)))